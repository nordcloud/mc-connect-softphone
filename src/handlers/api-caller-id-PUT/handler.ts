import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { StatusCodes } from 'http-status-codes';

import { getUser } from '../../utils/getUser';
import { putCallerId } from './putCallerId';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return { statusCode: StatusCodes.UNAUTHORIZED };
  }

  const callerId = event.body || '';
  const { expiresTimestampSeconds } = await putCallerId({ user, callerId });

  return {
    statusCode: StatusCodes.OK,
    body: JSON.stringify({ expiresTimestampSeconds }),
  };
};
