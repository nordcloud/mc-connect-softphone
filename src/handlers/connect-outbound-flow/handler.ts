import { ConnectContactFlowHandler } from 'aws-lambda';

import { deleteCallerId } from '../../utils/deleteCallerId';

export const handler: ConnectContactFlowHandler = async (event) => {
  const username = event.Details.Parameters.Username;

  if (!username) {
    throw new Error('Missing Username');
  }

  const callerId = await deleteCallerId(username);

  return {
    caller_id: callerId || null,
  };
};
