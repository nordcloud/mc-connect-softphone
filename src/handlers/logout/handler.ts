import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getCookie } from '../../utils/getCookie';
import { getStageConsts } from '../../utils/getStageConsts';

const { OAUTH_BASE_URL } = getStageConsts();

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { API_URL } = process.env;

  if (!API_URL) {
    throw new Error('Missing API_URL');
  }

  const idToken = getCookie(event, ID_TOKEN_COOKIE);
  let location = `${OAUTH_BASE_URL}/logout?id_token_hint=${idToken}&post_logout_redirect_uri=${API_URL}`;

  if (!idToken) {
    console.log('Missing idToken cookie');
    location = '/';
  }

  console.log(`Redirecting to ${location}`);

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location },
    cookies: [cookie.serialize(ID_TOKEN_COOKIE, '')],
  };
};
