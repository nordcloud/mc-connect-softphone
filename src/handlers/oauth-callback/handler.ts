import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getStage } from '../../utils/getStage';
import { fetchIdToken } from './fetchIdToken';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const responseCookies = [];
  const idTokenData = await fetchIdToken(event);

  if (idTokenData) {
    responseCookies.push(
      cookie.serialize(ID_TOKEN_COOKIE, idTokenData.id_token, {
        secure: getStage() !== 'local',
        httpOnly: true,
        expires: new Date(Date.now() + idTokenData.expires_in * 1000),
      })
    );
  }

  console.log('Redirecting to "/"');

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: '/' },
    cookies: responseCookies,
  };
};
