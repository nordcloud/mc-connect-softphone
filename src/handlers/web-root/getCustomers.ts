import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { CUSTOMERS_TABLE_ARN } from '../../consts';
import { customersMock } from '../../mockData';
import { CustomersTableItem } from '../../types';
import { getStage } from '../../utils/getStage';
import { scanDynamoDB } from '../../utils/scanDynamoDB';

export async function getCustomers(event: APIGatewayProxyEventV2) {
  const isMockMode =
    getStage() === 'local' &&
    Object.keys(Object(event.queryStringParameters)).includes('mock_db');

  if (isMockMode) {
    return customersMock;
  }

  const tableName = CUSTOMERS_TABLE_ARN.split('/').at(-1);

  if (!tableName) {
    throw new Error('Missing customers table name');
  }

  const items = await scanDynamoDB<CustomersTableItem>({ tableName });
  const sortedItems = items.sort((a, b) => a.CustomerName.localeCompare(b.CustomerName));

  return sortedItems;
}
