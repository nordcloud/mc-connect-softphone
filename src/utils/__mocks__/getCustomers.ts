export const getCustomers = jest.fn().mockResolvedValue({
  CustomerId: 'dummyCustomerId',
  CustomerName: 'dummyCustomerName',
  OutboundNumber: 'dummyOutboundNumber',
});
