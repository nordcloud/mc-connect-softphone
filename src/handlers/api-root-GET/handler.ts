import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { deleteCallerId } from '../../utils/deleteCallerId';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  await deleteCallerId(user.email);

  const customers = await getCustomers();

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers,
  });
};
