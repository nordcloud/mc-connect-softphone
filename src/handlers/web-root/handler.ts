import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers: await getCustomers(event),
  });
};
