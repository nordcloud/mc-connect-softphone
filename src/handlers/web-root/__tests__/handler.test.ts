import { getCustomers } from '../../../utils/getCustomers';
import { getUser } from '../../../utils/getUser';
import { loginRedirect } from '../../../utils/loginRedirect';
import { renderHTML } from '../../../utils/renderHTML';
import { handler } from '../handler';

jest.mock('../../../utils/getCustomers');
jest.mock('../../../utils/getUser');
jest.mock('../../../utils/loginRedirect');
jest.mock('../../../utils/renderHTML');

const args = ['dummyEvent'] as unknown as Parameters<typeof handler>;

describe('web-root handler', () => {
  it('calls getUser with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(getUser).mock.calls).toMatchSnapshot();
  });

  it('calls renderHTML with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(renderHTML).mock.calls).toMatchSnapshot();
  });

  it('calls getCustomers with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(getCustomers).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  describe('when user is undefined', () => {
    beforeEach(() => {
      jest.mocked(getUser).mockResolvedValueOnce(undefined);
    });

    it('calls loginRedirect with correct args', async () => {
      await handler(...args);
      expect(jest.mocked(loginRedirect).mock.calls).toMatchSnapshot();
    });

    it('resolves with a correct value', () => {
      return expect(handler(...args)).resolves.toMatchSnapshot();
    });
  });
});
