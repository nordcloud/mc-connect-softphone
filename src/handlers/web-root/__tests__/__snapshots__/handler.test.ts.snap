// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web-root handler calls getCustomers with correct args 1`] = `
[
  [
    "dummyEvent",
  ],
]
`;

exports[`web-root handler calls getUser with correct args 1`] = `
[
  [
    "dummyEvent",
  ],
]
`;

exports[`web-root handler calls renderHTML with correct args 1`] = `
[
  [
    "homepage",
    {
      "AWS_CONNECT_URL": "https://nordcloud-connect.my.connect.aws/ccp-v2/softphone",
      "customers": {
        "CustomerId": "dummyCustomerId",
        "CustomerName": "dummyCustomerName",
        "OutboundNumber": "dummyOutboundNumber",
      },
      "user": {
        "email": "dummyEmail",
        "idToken": "dummyIdToken",
        "name": "dummyName",
        "sub": "dummySub",
      },
    },
  ],
]
`;

exports[`web-root handler resolves with a correct value 1`] = `"dummyHTML"`;

exports[`web-root handler when user is undefined calls loginRedirect with correct args 1`] = `
[
  [],
]
`;

exports[`web-root handler when user is undefined resolves with a correct value 1`] = `"dummyLoginRedirect"`;
